engine/dlib/src/dlib/uri.cpp

```diff
@@ -125,38 +125,45 @@ namespace dmURI
         return false;
     }
 
-    void Encode(const char* src, char* dst, uint32_t dst_len)
+    Result Encode(const char* src, char* dst, uint32_t dst_len, uint32_t* bytes_written)
     {
         assert(src != (const char*) dst);
-        assert(dst_len > 0);
+        assert(dst == 0 || dst_len > 0);
 
-        uint32_t dst_left = dst_len - 1; // Make room for null termination
+        uint32_t dst_left = dst == 0 ? 0xFFFFFFFF : dst_len - 1; // Make room for null termination
         const char* s = src;
         char* d = dst;
         while (*s) {
             char c = *s;
             if (IsUnreserved(c)) {
                 if (dst_left >= 1) {
-                    *d = c;
+                    if (dst)
+                        *d = c;
                     s++;
                     d++;
                     dst_left--;
                 } else {
-                    break;
+                    *d = '\0';
+                    return RESULT_TOO_SMALL_BUFFER;
                 }
             } else {
                 if (dst_left >= 3) {
-                    dmSnPrintf(d, 4, "%%%02X", c);
+                    if (dst)
+                        dmSnPrintf(d, 4, "%%%02X", c);
                     s++;
                     d += 3;
                     dst_left -= 3;
                 } else {
-                    break;
+                    *d = '\0';
+                    return RESULT_TOO_SMALL_BUFFER;
                 }
             }
         }
-
-        *d = '\0';
+        if (dst)
+            *d = '\0';
+        if (bytes_written)
+            *bytes_written = d - dst + 1;
+        return RESULT_OK;
```

Comment:
It felt weird to not have a return value, and that one couldn't determine the required buffer size.